File: CPP/Petshop.cpp
#include <bits/stdc++.h>

using namespace std;

class Petshop {
private:
    int id;
    string nama;
    string kategori;
    int harga;
    Petshop* next; // Pointer untuk linked list

public:
    // Constructor 
    Petshop() {
        this->id = 0;
        this->nama = "";
        this->kategori = "";
        this->harga = 0;
        this->next = nullptr;
    }

    // Constructor dengan Parameter
    Petshop(int id, string nama, string kategori, int harga) {
        this->id = id;
        this->nama = nama;
        this->kategori = kategori;
        this->harga = harga;
        this->next = nullptr;
    }

    // Getter dan Setter
    int get_id() {
		return this->id; 
	}
	
    void set_id(int id) {
		this->id = id; 
	}

    string get_nama() {
		return this->nama; 
	}
	
    void set_nama(string nama) {
		this->nama = nama; 
	}

    string get_kategori() {
		return this->kategori; 
	}
	
    void set_kategori(string kategori) {
		this->kategori = kategori; 
	}

    int get_harga() {
		return this->harga; 
	}
	
    void set_harga(int harga) {
		this->harga = harga; 
	}

    Petshop* get_next() {
		return this->next; 
	}
	
    void set_next(Petshop* next) {
		this->next = next; 
	}

    // Menampilkan Peralatan yang ada di list
    void tampilkanPeralatan() {
        Petshop* temp = this;
        if (temp == nullptr) {
            cout << "Tidak ada Peralatan di dalam Petshop.\n";
            return;
        }

        cout << "Daftar Peralatan Petshop:\n";
        while (temp != nullptr) {
            cout << "ID: " << temp->get_id()
                 << ", Nama Produk: " << temp->get_nama()
                 << ", Kategori Produk: " << temp->get_kategori()
                 << ", Harga Produk: " << temp->get_harga() << endl;
            temp = temp->get_next();
        }
    }

    // Menambahkan Peralatan ke dalam list
    void tambahPeralatan(Petshop*& head, int& nextID, string nama, string kategori, int harga) {
        Petshop* newPeralatan = new Petshop(nextID++, nama, kategori, harga);

        if (head == nullptr) {
            head = newPeralatan;
			cout << "Peralatan berhasil ditambahkan.\n";
        } else {
            Petshop* temp = head;
            while (temp->get_next() != nullptr) {
                temp = temp->get_next();
            }
            temp->set_next(newPeralatan);
			cout << "Peralatan berhasil ditambahkan.\n";
        }
    }

    // Mengubah data Peralatan
    void ubahPeralatan(int id, string nama, string kategori, int harga) {
        Petshop* temp = this;
        while (temp != nullptr) {
            if (temp->get_id() == id) {
                temp->set_nama(nama);
                temp->set_kategori(kategori);
                temp->set_harga(harga);
				cout << "Peralatan berhasil diperbarui.\n";
                return;
            }
            temp = temp->get_next();
        }
        cout << "Peralatan dengan ID " << id << " tidak ditemukan.\n";
    }

    // Menghapus Peralatan dari list
    void hapusPeralatan(Petshop*& head, int id) {
        if (head == nullptr) {
            cout << "Tidak ada Peralatan untuk dihapus.\n";
            return;
        }

        if (head->get_id() == id) {
            Petshop* temp = head;
            head = head->get_next();
            delete temp;
            return;
        }

        Petshop* temp = head;
        while (temp->get_next() != nullptr && temp->get_next()->get_id() != id) {
            temp = temp->get_next();
        }

        if (temp->get_next() == nullptr) {
            cout << "Peralatan dengan ID " << id << " tidak ditemukan.\n";
            return;
        }

        Petshop* delNode = temp->get_next();
        temp->set_next(delNode->get_next());
        delete delNode;
        cout << "Peralatan berhasil dihapus.\n";
    }

    // Mencari Peralatan berdasarkan nama
    void cariPeralatan(string nama) {
        Petshop* temp = this;
        bool found = false;

        while (temp != nullptr) {
            if (temp->get_nama() == nama) {
                cout << "Peralatan ditemukan:\n";
                cout << "ID: " << temp->get_id()
                     << ", Nama Produk: " << temp->get_nama()
                     << ", Kategori Produk: " << temp->get_kategori()
                     << ", Harga Produk: " << temp->get_harga() << endl;
                found = true;
            }
            temp = temp->get_next();
        }

        if (!found) {
            cout << "Peralatan dengan nama '" << nama << "' tidak ditemukan.\n";
        }
    }

    // Destructor
    ~Petshop() {}
};

File: PHP/KODE PROGRAM/Petshop.php
<?php
class Petshop {
    private $id;        
    private $nama; 
    private $kategori; 
    private $harga;    
    private $foto;     
    private $next;      // Pointer ke peralatan berikutnya (Linked List)

    // Konstruktor untuk inisialisasi objek peralatan petshop
    public function __construct($id, $nama, $kategori, $harga, $foto) {
        $this->id = $id;
        $this->nama = $nama;
        $this->kategori = $kategori;
        $this->harga = $harga;
        $this->foto = $foto;
        $this->next = null;  // Awalnya tidak ada elemen berikutnya
    }

    // Getter
    public function getId() {
        return $this->id;
    }
    public function getNama() {
        return $this->nama;
    }
    public function getKategori() {
        return $this->kategori;
    }
    public function getHarga() {
        return $this->harga;
    }
    public function getFoto() {
        return $this->foto;
    }
    public function getNext() {
        return $this->next;
    }

    // Setter
    public function setNama($nama) {
        $this->nama = $nama;
    }
    public function setKategori($kategori) {
        $this->kategori = $kategori;
    }
    public function setHarga($harga) {
        $this->harga = $harga;
    }
    public function setFoto($foto) {
        $this->foto = $foto;
    }
    public function setNext($next) {
        $this->next = $next;
    }

    // MENAMPILKAN PERALATAN 
    public function tampilkanPeralatan($cariNama = "") {
        $temp = $this;  // Mulai dari kepala linked list
        $found = false;
        echo "<h4>Hasil Pencarian:</h4><ul>";

        // Loop untuk menelusuri semua peralatan dalam linked list
        while ($temp !== null) {
            // Jika pencarian kosong atau nama cocok dengan input user
            if ($cariNama === "" || stripos($temp->getNama(), $cariNama) !== false) {
                $found = true;
                echo "<li>ID: " . $temp->getId() . 
                    ", Nama: " . $temp->getNama() . 
                    ", Kategori: " . $temp->getKategori() . 
                    ", Harga: Rp " . number_format($temp->getHarga()) . 
                    "<br><img src='uploads/" . $temp->getFoto() . "' width='100'><br>
                    <a href='?hapus=" . $temp->getId() . "'>Hapus</a> |
                    <a href='?ubah=" . $temp->getId() . "'>Ubah</a></li><br>";
            }
            $temp = $temp->getNext();  // Pindah ke elemen berikutnya
        }

        // Jika tidak ada yang ditemukan
        if (!$found) {
            echo "<p>Data tidak ditemukan.</p>";
        }

        echo "</ul>";
    }

    //  MENAMBAH PERALATAN 
    public function tambahPeralatan($id, $nama, $kategori, $harga, $foto) {
        $newPeralatan = new Petshop($id, $nama, $kategori, $harga, $foto); // Buat objek baru
        $temp = $this;
        
        // Cari elemen terakhir dalam linked list
        while ($temp->getNext() !== null) {
            $temp = $temp->getNext();
        }
        $temp->setNext($newPeralatan); // Tambahkan peralatan baru di akhir linked list
    }

    //  MENGHAPUS PERALATAN 
    public function hapusPeralatan($id) {
        // Jika elemen pertama yang ingin dihapus
        if ($this->id == $id) {
            return $this->next; // Mengembalikan node berikutnya sebagai head baru
        }

        $temp = $this;

        // Mencari node sebelum yang ingin dihapus
        while ($temp->getNext() !== null) {
            if ($temp->getNext()->getId() == $id) {
                $temp->setNext($temp->getNext()->getNext()); // Lewati node yang akan dihapus
                return $this;
            }
            $temp = $temp->getNext();
        }
        return $this;
    }

    //  MENGUBAH DATA PERALATAN 
    public function ubahPeralatan($id, $nama, $kategori, $harga, $foto) {
        $temp = $this; // Mulai dari head linked list

        // Cari peralatan berdasarkan ID
        while ($temp !== null) {
            if ($temp->getId() == $id) {
                $temp->setNama($nama);
                $temp->setKategori($kategori);
                $temp->setHarga($harga);
                
                // Jika foto baru ada, update foto
                if ($foto !== null && $foto !== "") { 
                    $temp->setFoto($foto);
                }
                break;
            }
            $temp = $temp->getNext(); // Pindah ke peralatan berikutnya
        }
    }  
}
?>
